designs/0032-stan-output-formats.md

```diff
@@ -0,0 +1,336 @@
+   + `metric_type` - One of `unit`, `diag`, `dense`.
+
+- `model_metatdata` - A hierarchical object which contains the model variable names, types, dimensions, and declaration block.
+- `config` - A hierarchical object which contains the algorithm configuration, model name, and inference type.
+- `timestamps` - A table containing pairs of timestamps, event descriptors one row per event.
+
+Optional outputs (under the current interfaces):
+
+- `sample_params_unconstrained` - A table containing the per-draw estimates of just the model parameter and transformed parameters on the unconstrained scale; 1 column per variable (element), one row per iteration.
+
+Outputs not currently available:
+
+- `algorithm_internal_state` - A table which provides finer-grained reporting of the inference engine state for development, testing, and debugging purposes, one row per operation, e.g., leapfrog step.
+
+### Output formats
+
+JSON notation will be used for hierarchical and structured data.
+Tabular data will be either in
+
+- CSV format (human-readable)
+
+- Apache Arrow format, which is a self-describing binary data format.
+An Apache Arrow parquet file consists of an schema followed by one or more rows of data.
+The schema describes the structure of the data objects in each row, allowing the downstream
+process to reconstitute structured objects properly.
+The Arrow libraries for R and Python will allow downstream analysis packages
+to process these files, either during inference or afterwards.
+
+- Unformatted - the binary values returned by methods on the algorithm and model;
+all responsibility for handling these outputs lies with the caller.
+
+The Stan CSV output format will remain available via CmdStan.
+
+### Input formats
+
+Outputs from one inference can become inputs to a subsequent one.
+The input formats used by the inference algorithms are:
+
+- data variable definitions in JSON or [rdump](https://mc-stan.org/docs/cmdstan-guide/rdump.html)
+- initial parameter variables in JSON or rdump
+- for the standalone generated quantities, inputs include the Stan CSV file which contains the sample from the fitted model.
+
+Information from one inference run which would become either data or parameter inits would require converting CSV files to JSON.
+For standalone generated quantities, a sample in Apache format would need to be converted to CSV format.
+We will use existing converter libraries to handle this.
+
+## Technical Specification
+
+To implement the features outlined in the Functional Specification
+will require changes in the `stan::services` and `stan::callbacks` layers.
+
+We will define two [C++ Enumeration](https://en.cppreference.com/w/cpp/language/enum)
+classes: `InfoType` and `OutputFormat`.
+
+- The values of enum class `InfoType` correspond to the information items listed in
+the functional specification, e.g., `LP` (log_prob), `sample`, `metric`, etc.
```

Comment:
This is a prime area for discovery during implementation, but just for the sake of getting names right I think you'd probably want them to be `Lp`, `Draws` or `Samples`, and `Adaptation` (not just metric, e.g. the same info type could also be step size?)

Anyway, I think it's worth clarifying that the exact names and list of these will be part of the implementation, not something fixed in advance